[Trivial] BitFlags Clear function fails on type conversion

### Type

Other

### Explanation
https://github.com/project-chip/connectedhomeip/actions/runs/7008674355/job/19065470504?pr=30562

```
INFO    ../../examples/all-clusters-app/asr/third_party/connectedhomeip/src/lib/support/BitFlags.h: In instantiation of 'chip::BitFlags<FlagsEnum, StorageType>& chip::BitFlags<FlagsEnum, StorageType>::Clear(const chip::BitFlags<FlagsEnum, StorageType>&) [with FlagsEnum = chip::app::Clusters::BooleanSensorConfiguration::AlarmModeBitmap; StorageType = unsigned char]':
INFO    ../../examples/all-clusters-app/asr/third_party/connectedhomeip/src/app/clusters/boolean-sensor-configuration-server/boolean-sensor-configuration-server.cpp:212:29:   required from here
INFO    ../../examples/all-clusters-app/asr/third_party/connectedhomeip/src/lib/support/BitFlags.h:105:16: error: conversion from 'int' to 'unsigned char' may change value [-Werror=conversion]
INFO      105 |         mValue &= ~other.mValue;
INFO          |         ~~~~~~~^~~~~~~~~~~~~~~~
INFO    cc1plus: error: unrecognized command line option '-Wno-unknown-warning-option' [-Werror]
```

### Platform

all

### Platform Version(s)

8f76c06231aebbf5e5d09a00b7b5887c4a8f5bf6

### Testing

CI tested

### (Optional) If manually tested please explain why this is only manually tested

_No response_

### Anything else?

_No response_

--- a/src/lib/support/BitFlags.h
+++ b/src/lib/support/BitFlags.h
@@ -102,7 +102,7 @@ public:
      */
     BitFlags & Clear(const BitFlags & other)
     {
-        mValue &= ~other.mValue;
+        mValue &= static_cast<IntegerType>(~static_cast<IntegerType>(other.mValue));
         return *this;
     }
 

--- a/src/lib/support/tests/TestBitMask.cpp
+++ b/src/lib/support/tests/TestBitMask.cpp
@@ -98,10 +98,26 @@ void TestBitMaskInvalid(nlTestSuite * inSuite, void * inContext)
     NL_TEST_ASSERT(inSuite, mask.Raw() == 0x1234);
 }
 
+void TestClear(nlTestSuite * inSuite, void * inContext)
+{
+    BitMask<TestEnum> mask1;
+    BitMask<TestEnum> mask2;
+
+    mask1.Set(TestEnum::kBit_0);
+    mask1.Set(TestEnum::kBits_1_2);
+    mask1.Set(TestEnum::kBits_High8);
+
+    mask2.Set(TestEnum::kBits_1_2);
+    mask1.Clear(mask2);
+
+    NL_TEST_ASSERT(inSuite, mask1.Raw() == 0xFF01);
+}
+
 const nlTest sTests[] = {
     NL_TEST_DEF("BitMask operations", TestBitMaskOperations), //
     NL_TEST_DEF("BitFields logic", TestBitFieldLogic),        //
     NL_TEST_DEF("Invalid operations", TestBitMaskInvalid),    //
+    NL_TEST_DEF("Clear operations", TestClear),               //
     NL_TEST_SENTINEL()                                        //
 };
 

